### ISSUE TITLE
Format quickfix output 

### ISSUE THREAD

**Issue by ngortheone** [vim/vim#4425]
As @justinmk suggetsted I am re-opening https://github.com/neovim/neovim/issues/10056 here

I write rust with rls and often use quickfix window to find issues.

If you compare how similar "quickfix" window looks in Emacs you will notice that the output is
justified by columns, and overall is more readable.

Emacs flycheck:
![emacs-flycheck](https://user-images.githubusercontent.com/18271144/58334116-dc2aa380-7e0c-11e9-98cc-84f4aae4e587.png)

vim quickfix:
![vim-quickfix](https://user-images.githubusercontent.com/18271144/58334128-e2208480-7e0c-11e9-82fc-a6bbd4467695.png)

I have not found any options to format the output of the quickfix window. It would be great to have
the ability to provide printf-like format string for quickfix item, specifying column lengths, and inlucing/excluding items like column numbers, error codes etc.

**Comment by tonymec** [comment]
I often use quickfix commands, but I rarely use the quickfix window. I have in fact some of the following mappings for easier moving about in the quickfix list:
```vim
if exists(':map') == 2
  if has('quickfix')
    map <F2>   :cnext<CR>
    map <S-F2> :cprev<CR>
    map <F3>   :cnfile<CR>
    map <S-F3> :cpfile<CR>
    map <F4>   :cfirst<CR>
    map <S-F4> :clast<CR>
  endif
endif
```
Some are more useful than others so you may want to put only some of them in your vimrc.

This said, I **think** that what you see is due to the history of how the quickfix feature was added to Vim: the :make command and the 'errorformat' option were added first, to help Vim decode compilers' output messages and bring up the corresponding source line. Then (IIRC) :grep, :helpgrep and :vimgrep (in that order) were added to treat the grep utility as a compiler, and then to add Vim-intrinsic capabilities for searching the helpfiles, then any files, according to Vim's own regular expresion syntax.

The format of a compiler's output is of course something over which Vim has no control. In the case of the intrinsic :helpgrep and :vimgrep commands, there are pluses and minuses to what you are requesting: among the minus points, aligning the text would IIUC require two-pass handling, one pass to determine the longest string in each column and then a second pass to output the text; then the extra spaces would mean that the visible source text would be truncated earlier than now whenever the text representing the filename, line and column would be shorter than the maximum (and therefore be padded with a certain number of spaces).

Best regards,
Tony.

**Comment by lacygoill** [comment]
This is not what you want, but in the meantime, to make the entries aligned, I use something like this in `~/.vim/after/plugin/qf.vim`:

    if executable('column') && executable('sed')
        let s:ul_save = &l:ul
        setl modifiable ul=-1
        sil! exe "%!sed 's/|/\<c-a>|/1'"
        sil! exe "%!sed 's/|/\<c-a>|/2'"
        sil! exe "%!column -s '\<c-a>' -t"
        let &l:ul = s:ul_save
        unlet! s:ul_save
        setl nomodifiable nomodified
    endif

It requires the `$ sed` and `$ column` utilities.

And to shorten the file paths, if you have the patch [8.0.1782](https://github.com/vim/vim/commit/d76ce852668635d81778cedacc2d3f021ed4e475), you can use the `'module'` property of the quickfix list to change the text which is displayed:

    call setqflist([], 'r', {'items': map(getqflist(),
        \ {i,v -> extend(v, {'module': substitute(fnamemodify(bufname(v.bufnr), ':t'), '.\{7}\zs.\+', '…', '')})})})
    if executable('column') && executable('sed')
        let s:ul_save = &l:ul
        setl modifiable ul=-1
        sil! exe "%!sed 's/|/\<c-a>|/1'"
        sil! exe "%!sed 's/|/\<c-a>|/2'"
        sil! exe "%!column -s '\<c-a>' -t"
        let &l:ul = s:ul_save
        unlet! s:ul_save
        setl nomodifiable nomodified
    endif

**Comment by blueyed** [comment]
Just for reference: Neomake has an experimental/optional feature to re-format the quickfix/location list (via `neomake#quickfix#enable()`).
It applies formatting according to / based on the makers/linters that were used, and adds basic highlighting, e.g. for `E\d+` from Python's flake8.
It also shortens / aligns columns, and only displays the buffer name if necessary (when multiple buffers are in the list), otherwise (and in general) the quickfix title is used.
lnum/cols and maker names are aligned.

With multiple buffers it adds the name to the first entry where it changes.  When it would use a separate column for this, a lot of space would be wasted (since that's repeated information).

![2019-06-01-041634_952x515_escrotum](https://user-images.githubusercontent.com/9766/58742312-1820c280-8424-11e9-95b3-66bf22731f25.png)

I see that this is very specific to Neomake's use case, and especially for having a single buffer therein mostly, but just wanted to share this anyway

**Comment by blueyed** [comment]
One quick/good improvement would be to use `fname:39:15` instead of `fname|39 col 15|`, which also makes it easier to copy'n'paste for other tools (that handle `fname:lnum`).
Although it should probably not use `fname:col` then, but maybe rather `fname::col` (for then there is no lnum, but only a column).

**Comment by vim-ml** [comment]
Hi,

I have been thinking about implementing something along the lines of the
'statusline' option (e.g. 'qfline') for formatting the lines displayed in
the
quickfix buffer.

- Yegappan

**Comment by chrisbra** [comment]
isn't it possible to make use of concealing for aligning?

**Comment by vim-ml** [comment]
Hi,

> On Fri, May 31, 2019 at 10:35 PM Yegappan Lakshmanan <yegappanl@gmail.com>
wrote:
> Hi,
>
> I have been thinking about implementing something along the lines of
> the 'statusline' option (e.g. 'qfline') for formatting the lines
> displayed in the quickfix buffer.
>

I looked into implementing this option. Currently each line in the
quickfix buffer has three columns (filename, line/column/error type
information and error message). To align these columns, we can support
the following format for this option:

     set qfline=20,10

where, the first number specifies the width of the first column and the
second number specifies the width of the second column and the rest of
the line will be used for the third column. The values within the
columns will be right-aligned. So if a long filename is used, only the
rightmost characters in the filename will be displayed.

Any comments/suggestions on this?

I also considered supporting a more generic printf like format for this
option:

   set qfline=%20f,%10l,%10m

where the fields are optional and additional fields from the quickfix
entry can be used. I am not sure whether this level of flexibility
is needed for this option.

Thanks,
Yegappan

**Comment by ngortheone** [comment]
@vim-ml how can I try this option? Is this in master?

**Comment by vim-ml** [comment]
Hi,

On Sat, Jun 15, 2019 at 11:46 AM Ihor Antonov <vim-dev-github@256bit.org> wrote:
>
> @vim-ml how can I try this option? Is this in master?
>

No. This is not yet implemented and merged. I am looking for comments
before implementing it.

- Yegappan

**Comment by g0xA52A2A** [comment]
Just as an FYI there are plugins like [vim-addon-qf-layout](https://github.com/MarcWeber/vim-addon-qf-layout) that alter the contents of the quickfix window. I use [this](https://gist.github.com/george-b/d59600bde83dd4f7353aef8011315f2f) simplified version.

This will of course throw off what Vim considers to be the expected syntax. So if you care about that you'll need to re-define `qfFileName`, `qfSeparator`, `qfLineNr`, and `qfError` accordingly.

**Comment by lacygoill** [comment]
I think this is fixed by [8.2.0869](https://github.com/vim/vim/releases/tag/v8.2.0869) and [8.2.0959](https://github.com/vim/vim/releases/tag/v8.2.0959), which provide the global option `'quickfixtextfunc'`, as well as a quickfix list attribute with the same name.

---

I had a look at the screenshot; you could get the desired result, or at least something close enough with this code:

    vim -Nu NONE -S <(cat <<'EOF'
        set qftf=QuickFixTextFunc nowrap | syn on
        fu QuickFixTextFunc(info) abort
            if a:info.quickfix
                let qfl = getqflist(#{id: a:info.id, items: 0}).items
            else
                let qfl = getloclist(a:info.winid, #{id: a:info.id, items: 0}).items
            endif
            let l = []
            for idx in range(a:info.start_idx - 1, a:info.end_idx - 1)
                let e = qfl[idx]
                let fname = bufname(e.bufnr)->fnamemodify(':t')
                if strchars(fname, 1) <= 8
                    let fname = printf('%8s', fname)
                else
                    let fname = printf('%.7s', fname)..'…'
                endif
                let lnum = printf('%5d', e.lnum)
                let col = printf('%3d', e.col)
                call add(l, fname..'|'..lnum..' col '..col..'| '..e.text)
            endfor
            return l
        endfu
        helpg foobar
        copen
    EOF
    )

**Comment by chrisbra** [comment]
closing then. Thanks.
